### ISSUE TITLE
Feature Request: PostgreSQL Full Text Search support

### ISSUE THREAD

**Issue by timkofu** [django-haystack/django-haystack#1320]
Any plans to add Postgres as a back-end?

**Comment by thedrow** [comment]
There are plans to add it to `django.contrib.postgres` at some point.

**Comment by acdha** [comment]
I would prefer to have new backends be developed as separate apps (possibly hosted in this repo) following the lead of the Xapian backend. Python packaging has improved a lot since Haystack started and it seems like it'd be better to start actively shrinking our core text matrix.

**Comment by auvipy** [comment]
@acdha agreed. haystack is consisting much at a single project so better plan to split back ends out as possible and keep haystack as a common interface for them.

**Comment by auvipy** [comment]
django 1.10a1 supports PGSQL FTS

**Comment by SalahAdDin** [comment]
Interesting, but, i don't know if is a good idea split haystack in may smallest apps.

**Comment by acdha** [comment]
@auvipy That's a very good point – that might be a good pilot project, simply forking the existing simple backend to produce a Postgres-specific one with minimal overhead beyond core Django. Given the limited number of volunteers contributing to Haystack, I really like the idea of avoiding maintenance overhead here when Django offers an alternative (following the lead of deprecating the old generic views in favor of Django's standard ones).

**Comment by gfairchild** [comment]
Has there been any more thought/discussion on this issue? I'm very interested in using the built-in Django 1.10 pgSQL full-text capabilities over ElasticSearch on one of my projects. We're already using Haystack + ES, but migrating the backend to pgSQL full-text would be nice for several reasons.

**Comment by acdha** [comment]
@gfairchild I still think the best approach would be creating a separate project with a new backend based on the existing simple backend (we could host that repo under the main django-haystack organization, too). The hard part is finding people with the time to work on it but if you were interested I suspect it would be a relatively modest amount of work given how much Haystack has tried to follow Django conventions.

**Comment by snake-soft** [comment]
_Update-question after 4 years:_
Is there any approach to use postgres full text search with haystack?

**Comment by SalahAdDin** [comment]
The question is: does this package still work?

**Comment by auvipy** [comment]
> _Update-question after 4 years:_
> Is there any approach to use postgres full text search with haystack?

django have them built in now

**Comment by snake-soft** [comment]
> > _Update-question after 4 years:_
> > Is there any approach to use postgres full text search with haystack?
> 
> django have them built in now

I guess you meant this:
https://docs.djangoproject.com/en/3.0/ref/contrib/postgres/search/

There are projects (eg. django-oscar) that have django-haystack fully integrated. Maybe they could be served all with one django-haystack-postgres plugin.

Can someone estimate if it is a comprehensive task to adapt the interface of postgres to the haystack facettes?

**Comment by acdha** [comment]
Haystack has a “simple” backend which uses the ORM. I think it would be relatively straight forward to use that as a basis for a Postgres extension — i.e. faceting would remain ORM filters but text matching would use the newer Django full-text search support  but there'd be a question about how to handle compatibility. Depending on the amount of code, it could be a simple check for the ORM backend features which would basically fall back to `__icontains` when it can't use Django's `__search` operator:

https://github.com/django-haystack/django-haystack/blob/1cc82885e287fe33bb937047137da3d803072684/haystack/backends/simple_backend.py#L67

**Comment by mslinn** [comment]
@savelmtr created a [PR for PostgreSQL full-text search](https://github.com/ellmetha/django-machina/pull/187) for another project, `django-machina`. The PR was not accepted - I do not follow the reasoning for why that decision was made. There was never any question about the quality of the PR's code.

That PR could be very helpful for whomever would like to make a PostgreSQL full-text search addon for `django-haystack`.

Fortunately, there would be zero politics involved in porting the PR @savelmtr made for `django-machina` to  `django-haystack`. This is because the consensus for the best way to add support for additional data stores to `django-haystack` is that the [xapian-haystack project](https://github.com/notanumber/xapian-haystack) shoulld be cloned and modified to work with PostgreSQL. Perhaps one of the [PostgreSQL sponsors](https://www.postgresql.org/about/sponsors/) might take on the work for PostgreSQL, or perhaps @savelmtr might be interested in doing the work if asked nicely.

Disclaimer: I have never interacted with @savelmtr and I have never used `django-machina`. The idea of incorporating the work that @savelmtr did on the PR into django-haystack just seems to make sense. I also have limited interaction with `django-haystack` - you can read about why I care [here](https://www.mslinn.com/blog/2021/02/14/django-fulltext-search.html#suggestionsHay).

**Comment by savelmtr** [comment]
Ok, people, now I've packed this search in a module: https://github.com/savelmtr/django-machina-search

**Comment by manelclos** [comment]
Hi all, I've looking at the alternatives explained here but found none that implements Haystack's search backend interface. What we need is a drop-in replacement for the whoosh backend that has served us so well for many years, but is suffering performance problems since we added multiple conditions to the search. We have used PostgreSQL FTS directly in a different project, but still waiting for a haystack backend that would allow us to keep the search indexes and related code with no changes, i.e. just changing the Haystack configuration and reindexing.

We also value the flexibility of using PostgreSQL FTS through Haystack high level interface.

I'll give it a try. In the meantime, if anyone knows about any development to fix this issue, please let me know.

**Comment by savelmtr** [comment]
@mslinn, @manelclos it's about postgres support for django haystack. Of course, I could try to make such a module, it's not a rocket science. But honestly I'm not so sure if it is useful at all and - second doubt - if haystack supports adding new backends easily (as I remember, the part of integration is a true headache).

**Comment by manelclos** [comment]
@savelmtr I already have some code working, I need to add missing pieces and do a cleanup, then I will release it on github. On my side it is about compatibility with existing code, i.e. migrating from whoosh to Postgres FTS without any code change. same goes for people using Oscar I think. Also, it is a good thing to reuse haystack concepts with Postgres without reinventing the wheel.

**Comment by benzkji** [comment]
@manelclos is there a beta, a repos we could look at?

**Comment by fgregg** [comment]
i wrote a postgres FTS backend that is working pretty well for me: 

https://github.com/fgregg/postgres-fts-backend

**Comment by acdha** [comment]
> i wrote a postgres FTS backend that is working pretty well for me:
> 
> https://github.com/fgregg/postgres-fts-backend

Feel free to send a pull-request updating https://github.com/django-haystack/django-haystack/blob/master/docs/backend_support.rst with a link
